scripts/upgrade.ts

```diff
@@ -0,0 +1,254 @@
+import { diffLines } from "https://esm.sh/diff@5.1.0";
+import deno from "../deno.json" assert { type: "json" };
+import meta from "../meta.json" assert { type: "json" };
+import { namespaceFromImportMap } from "../utils/namespace.ts";
+
+const exists = async (dir: string): Promise<boolean> => {
+  try {
+    await Deno.stat(dir);
+    // successful, file or directory must exist
+    return true;
+  } catch (error) {
+    if (error instanceof Deno.errors.NotFound) {
+      // file or directory does not exist
+      return false;
+    } else {
+      // unexpected error, maybe permissions, pass it along
+      throw error;
+    }
+  }
+};
+
+interface UpgradeOption {
+  version?: string; // if omitted, latest will be used.
+  isEligible: () => Promise<boolean>;
+  apply(): Promise<FileMod[]>;
+}
+
+// v1 migration code
+interface Patch {
+  from: {
+    path: string;
+    content: string;
+  };
+  to: {
+    path: string;
+    content: string;
+  };
+}
+interface Delete {
+  path: string;
+}
+type FileMod = Patch | Delete;
+class UpgradeError extends Error {
+  constructor(err: string) {
+    super(err);
+  }
+}
+const updateImportMap = async (
+  liveVersion: string,
+  stdVersion: string,
+): Promise<Patch> => {
+  const importMapFile = (deno.importMap ?? "./import_map.json").replace(
+    "./",
+    "",
+  );
+
+  const importMapPath = join(Deno.cwd(), importMapFile);
+  if (!(await exists(importMapPath))) {
+    throw new UpgradeError(
+      `${importMapPath} is required to upgrade live dependency verson`,
+    );
+  }
+  const importMapStr = await Deno.readTextFile(importMapPath);
+  const { imports, ...rest }: { imports: Record<string, string> } = JSON.parse(
+    importMapStr,
+  );
+  return {
+    from: {
+      path: importMapPath,
+      content: importMapStr,
+    },
+    to: {
+      path: importMapPath,
+      content: JSON.stringify(
+        {
+          ...rest,
+          imports: {
+            ...imports,
+            "$live/": `https://denopkg.com/deco-cx/live@${liveVersion}/`,
+            "deco-sites/std/":
+              `https://denopkg.com/deco-sites/std@${stdVersion}/`,
+          },
+        },
+        null,
+        2,
+      ),
+    },
+  };
+};
+
+const siteId = async (): Promise<number | undefined> => {
+  const middleware = join(Deno.cwd(), "routes", "_middleware.ts");
+  if (!(await exists(middleware))) {
+    return undefined;
+  }
+  const reg = /siteId: (?<siteId>\d+)/gm;
+  const match = reg.exec(await Deno.readTextFile(middleware));
+  if (!match?.groups) {
+    return undefined;
+  }
+  return +match.groups["siteId"];
+};
+
+const updateDevTsImports = async () => {
+  const devTs = join(Deno.cwd(), "dev.ts");
+  if (!(await exists(devTs))) {
+    throw new UpgradeError(`${devTs} not found`);
+  }
+  const devTsContent = await Deno.readTextFile(devTs);
+  const updateFreshGen = devTsContent.replaceAll(
+    "fresh.gen.ts",
+    "live.gen.ts",
+  );
+  const id = await siteId();
```

Comment:
Really cool!

WDYT about placing `v0-to-v1.ts` in another file and having also a simple way to declare future upgrades (that might only be a version bump)?

Reply:
the way it is implemented already support to split in many files but I didn't do that for now, can we defer a bit this implementation?